compiler/src/dotty/tools/dotc/transform/DropBreaks.scala

```diff
@@ -0,0 +1,213 @@
+   *  `LabelsShadowedByTry` context property. This means that breaks to these
+   *  labels will not be translated to labeled returns in the body of the try.
+   */
+  override def prepareForTry(tree: Try)(using Context): Context = tree match
+    case LabelTry(_, _) => ctx
+    case _ => ctx.property(LabelUsages) match
+      case Some(usesMap) =>
+        val setSoFar = ctx.property(LabelsShadowedByTry).getOrElse(Set.empty)
+        ctx.fresh.setProperty(LabelsShadowedByTry, setSoFar ++ usesMap.keysIterator)
+      case _ => ctx
+
+  /** If `tree` is a BreakBoundary, transform it as follows:
+   *   - Wrap it in a labeled block if its label has local uses
+   *   - Drop the try/catch if its label has no other uses
+   */
+  override def transformBlock(tree: Block)(using Context): Tree = tree match
+    case BreakBoundary(label, expr) =>
+      val uses = ctx.property(LabelUsages).get(label)
+      val tree1 =
+        if uses.otherRefs > 1 then
+          // one non-local ref is always in the catch clause; this one does not count
+          tree
+        else
+          expr
+      report.log(i"trans boundary block $label // ${uses.returnRefs}, ${uses.otherRefs}")
+      if uses.returnRefs > 0 then Labeled(uses.goto, tree1) else tree1
+    case _ =>
+      tree
+
+  /** Rewrite a BreakThrow
+   *
+   *     val Label_this: ... = local
+   *     throw new Break[...](Label_this, arg)
+   *
+   *  where `local` is defined in the current method and is not included in
+   *  LabeldShowedByTry to
+   *
+   *     return[target] arg
+   *
+   *  where `target` is the `goto` return label associated with `local`.
+   *  Adjust associated ref counts accordingly. The local refcount is increased
+   *  and the non-local refcount is decreased, since `local` the `Label_this`
+   *  binding containing `local` is dropped.
+   */
+  override def transformInlined(tree: Inlined)(using Context): Tree = tree match
+    case BreakThrow(lbl, arg) =>
+      report.log(i"trans inlined $arg, ${arg.source}, ${ctx.outer.source}, ${tree.source}")
+      labelUsage(lbl) match
+        case Some(uses: LabelUsage)
+        if uses.enclMeth == ctx.owner.enclosingMethod
+            && !ctx.property(LabelsShadowedByTry).getOrElse(Set.empty).contains(lbl)
+          =>
+          uses.otherRefs -= 1
+          uses.returnRefs += 1
+          cpy.Inlined(tree)(tree.call, Nil,
+            inContext(ctx.withSource(tree.expansion.source)) {
+              Return(arg, ref(uses.goto)).withSpan(arg.span)
+            })
+        case _ =>
+          tree
+    case _ =>
+      tree
+
+  /** If `tree` refers to an enclosing label, increase its non local recount.
+   *  This increase is corrected in `transformInlined` if the reference turns
+   *  out to be part of a BreakThrow to a local, non-shadowed label.
+   */
+  override def transformIdent(tree: Ident)(using Context): Tree =
+    if tree.symbol.name == nme.local then
```

Comment:
Maybe we should check that `tree.tpe.derivesFrom(defn.LabelClass)` to not count all `Ident`s named `local`.

Reply:
The following labelUsage lookup is probably as cheap or cheaper than the derivesFrom test.